packages/core/src/common/telemetry.test.ts

```diff
@@ -10,30 +13,89 @@ beforeEach(() => {
   return () => vi.unstubAllGlobals();
 });
 
-test("events are processed", async (context) => {
-  const options = { ...context.common.options, telemetryDisabled: false };
-  const telemetry = new TelemetryService({ options });
+beforeEach((context) => {
+  const tempDir = path.join(os.tmpdir(), randomUUID());
+  mkdirSync(tempDir, { recursive: true });
+
+  context.common.options = {
+    ...context.common.options,
+    telemetryDisabled: false,
+    telemetryConfigDir: tempDir,
+  };
+
+  return async () => {
+    rmSync(tempDir, { force: true, recursive: true });
+  };
+});
+
+test("telemetry without existing conf creates new conf", async (context) => {
+  const telemetry = createTelemetry({
+    options: context.common.options,
+    logger: context.common.logger,
+  });
+
+  telemetry.record({
+    name: "lifecycle:heartbeat_send",
+    properties: { duration_seconds: process.uptime() },
+  });
 
-  telemetry.record({ event: "test", properties: { test: "data" } });
-  await telemetry.flush();
+  await telemetry.kill();
 
-  expect(fetchSpy).toHaveBeenCalled();
+  expect(fetchSpy).toHaveBeenCalledTimes(1);
+
+  const fetchUrl = fetchSpy.mock.calls[0][0];
+  expect(fetchUrl).toBe(context.common.options.telemetryUrl);
 
   const fetchBody = JSON.parse(fetchSpy.mock.calls[0][1].body);
   expect(fetchBody).toMatchObject({
-    anonymousId: expect.any(String),
-    context: expect.anything(),
-    event: "test",
-    properties: { test: "data" },
+    distinctId: expect.any(String),
+    event: "lifecycle:heartbeat_send",
+    properties: { duration_seconds: expect.any(Number) },
+  });
+});
+
+test("telemetry does not submit events if telemetry is disabled", async (context) => {
+  const telemetry = createTelemetry({
+    options: { ...context.common.options, telemetryDisabled: true },
+    logger: context.common.logger,
+  });
+
+  telemetry.record({
+    name: "lifecycle:session_start",
+    properties: { cli_command: "test" },
+  });
+  telemetry.record({
+    name: "lifecycle:heartbeat_send",
+    properties: { duration_seconds: process.uptime() },
   });
+
+  await telemetry.kill();
+
+  expect(fetchSpy).toHaveBeenCalledTimes(0);
 });
 
-test("events are not processed if telemetry is disabled", async (context) => {
-  const options = { ...context.common.options, telemetryDisabled: true };
-  const telemetry = new TelemetryService({ options });
+test("telemetry kill times out after 1 second and cancels pending requests", async (context) => {
+  const telemetry = createTelemetry({
+    options: context.common.options,
+    logger: context.common.logger,
+  });
+
+  for (let i = 0; i < 100; i++) {
+    telemetry.record({
+      name: "lifecycle:heartbeat_send",
+      properties: { duration_seconds: process.uptime() },
+    });
+  }
+
+  // Mock fetch to take 91ms to complete.
+  fetchSpy.mockImplementation(
+    () => new Promise((resolve) => setTimeout(resolve, 95)),
+  );
```

Comment:
Why 91ms? Also doesn't match the number provider to setTimeout

Reply:
Oh, these can both be 100 actually now that we're doing >8 and <12 below. I previously had it expecting 10 which appeared flaky.